R/loo_predictive_error.R

```diff
@@ -0,0 +1,200 @@
+#' mean squared error for continuous predictions and accuracy and balanced
+#' accuracy for binary classification. Predictions are passed on to the
+#' `E_loo()` function, so this function assumes that the PSIS approximation is
+#' working well.
+#'
+#' @param x A numeric matrix of predictions.
+#' @param y A numeric vector of observations. Length should be equal to the
+#'     number of rows in `x`.
+#' @param ll A matrix of pointwise log-likelihoods. Should be of same dimension
+#'     as `x`.
+#' @param pred_error The type of predictive error to be used. Currently
+#'     supported options are `"mae"`, `"rmse"` and `"mse"` for regression and
+#'     for binary classification `"acc"` and `"balanced_acc"`.
+#'     \describe{
+#'       \item{`"mae"`}{
+#'          Mean absolute error.
+#'       }
+#'       \item{`"mse"`}{
+#'          Mean squared error.
+#'       }
+#'       \item{`"rmse"`}{
+#'          Root mean squared error, given by as the square root of `MSE`.
+#'       }
+#'       \item{`"acc"`}{
+#'          The proportion of predictions indicating the correct outcome.
+#'       }
+#'       \item{`"balanced_acc"`}{
+#'          Balanced accuracy is given by the average of true positive and true
+#'          negative rates.
+#'       }
+#'     }
+#' @param r_eff A Vector of relative effective sample size estimates containing
+#'     one element per observation. See [psis()] for more details.
+#' @param ... Additional arguments passed on to [E_loo()]
+#'
+#' @return A list with the following components:
+#' \describe{
+#'   \item{`estimate`}{
+#'   Estimate of the given predictive error measure.
+#'   }
+#'  \item{`se`}{
+#'   Standard error of the estimate.
+#'   }
+#'  }
+#' @export
+#'
+#' @examples
+#' \donttest{
+#' if (requireNamespace("rstanarm", quietly = TRUE)) {
+#' # Use rstanarm package to quickly fit a model and get both a log-likelihood
+#' # matrix and draws from the posterior predictive distribution
+#' library("rstanarm")
+#'
+#' # data from help("lm")
+#' d <- data.frame(
+#'   weight = c(ctl, trt),
+#'   group = gl(2, 10, 20, labels = c("Ctl","Trt"))
+#' )
+#' fit <- stan_glm(weight ~ group, data = d, refresh = 0)
+#' ll <- log_lik(fit)
+#' r_eff <- relative_eff(exp(-log_ratios), chain_id = rep(1:4, each = 1000))
+#'
+#' mu_pred <- posterior_epred(fit)
+#' # Leave-one-out mean absolute error of predictions
+#' mae <- loo_predictive_error(x = mu_pred, y = d$weight, ll = ll,
+#'                             pred_error = 'mae', r_eff = r_eff)
+#' # Leave-one-out 90%-quantile of mean absolute error
+#' mae_90q <- loo_predictive_error(x = mu_pred, y = d$weight, ll = ll,
+#'                                 pred_error = 'mae', r_eff = r_eff,
+#'                                 type = 'quantile', probs = 0.9)
+#' }
+#' }
+loo_predictive_error <- function(x, ...) {
+  UseMethod("loo_predictive_error")
+}
+
+#' @rdname loo_predictive_error
+#' @export
+loo_predictive_error.default <-
+  function(x,
+           y,
+           ll,
```

Comment:
I wonder if we should call the argument `log_lik` instead of `ll` since for `psis()` we have an argument `log_ratios` and not `lr`. Not a big deal either way. @avehtari What do you think?

Reply:
longer name is more descriptive, and thus better

Reply:
Ok let's go with `log_lik`. @LeeviLindgren Can you update the argument name?